src/Heap.sol

```diff
@@ -32,209 +36,199 @@ library BasicHeap {
     /// @param _heap The heap to modify.
     /// @param _id The address of the account to decrease the amount.
     /// @param _newValue The new value of the account.
-    function decrease(
-        Heap storage _heap,
-        address _id,
-        uint256 _newValue
-    ) internal {
-        uint256 rank = _heap.ranks[_id];
-        if (rank == 0) revert AccountDoesNotExist();
-        uint256 oldValue = getAccount(_heap, rank).value;
-        if (_newValue >= oldValue || _newValue == 0) revert WrongValue();
-
-        setAccountValue(_heap, rank, _newValue);
-        shiftDown(_heap, rank);
+    function decrease(Heap storage _heap, address _id, uint256 _newValue) internal {
+        uint256 index = _heap.indexOf[_id];
+        if ((!containsAccount(_heap, index, _heap.accounts.length, _id))) revert AccountDoesNotExist();
+
+        uint256 oldValue = _heap.accounts[index].value;
+        if (_newValue >= oldValue) revert WrongValue();
+
+        shiftDown(_heap, _heap.accounts.length, Account(_id, _newValue), index);
     }
 
     /// @notice Increases the amount of an account in the `_heap`.
     /// @dev Only call this function when `_id` is in the `_heap` with a smaller value than `_newValue`.
     /// @param _heap The heap to modify.
     /// @param _id The address of the account to increase the amount.
     /// @param _newValue The new value of the account.
-    function increase(
-        Heap storage _heap,
-        address _id,
-        uint256 _newValue
-    ) internal {
-        uint256 rank = _heap.ranks[_id];
-        if (rank == 0) revert AccountDoesNotExist();
-        uint256 oldValue = getAccount(_heap, rank).value;
+    function increase(Heap storage _heap, address _id, uint256 _newValue) internal {
+        uint256 index = _heap.indexOf[_id];
+        if ((!containsAccount(_heap, index, _heap.accounts.length, _id))) revert AccountDoesNotExist();
+
+        uint256 oldValue = _heap.accounts[index].value;
         if (_newValue <= oldValue) revert WrongValue();
 
-        setAccountValue(_heap, rank, _newValue);
-        shiftUp(_heap, rank);
+        shiftUp(_heap, Account(_id, _newValue), index);
     }
 
     /// @notice Removes an account in the `_heap`.
     /// @dev Only call when `_id` is in the `_heap`.
     /// @param _heap The heap to modify.
     /// @param _id The address of the account to remove.
     function remove(Heap storage _heap, address _id) internal {
-        uint256 rank = _heap.ranks[_id];
-        if (rank == 0) revert AccountDoesNotExist();
-        uint256 removedValue = getAccount(_heap, rank).value;
-        uint256 size = getSize(_heap);
+        uint256 index = _heap.indexOf[_id];
+        uint256 accountsLength = _heap.accounts.length;
+
+        if ((!containsAccount(_heap, index, accountsLength, _id))) revert AccountDoesNotExist();
+
+        delete _heap.indexOf[_id];
 
-        if (rank == size) {
+        if (index == accountsLength - 1) {
```

Comment:
If we can make an assumption that accountsLength > 0, then this entire function can be unchecked.

Reply:
Yes it's safe to assume that `accountsLength > 0` because at this point we know that the account is in the heap.
I made the suggested modification, the code is more optimized but less readable